scipio/src/sys/rusage.rs

```diff
@@ -0,0 +1,123 @@
+#![allow(dead_code)]
+
+use std::ops::Sub;
+use std::time::Duration;
+
+#[derive(Debug, Clone)]
+pub struct ResourceUsage {
+    /// This is the total amount of time spent executing in user mode,
+    /// expressed in a timeval structure (seconds plus microseconds).
+    pub ru_utime: u64,
+    /// This is the total amount of time spent executing in kernel mode,
+    /// expressed in a timeval structure (seconds plus microseconds).
+    pub ru_stime: u64,
+    /// This is the maximum resident set size used (in kilobytes).
+    /// For RUSAGE_CHILDREN, this is the resident set size of the largest
+    /// child, not the maximum resident set size of the process tree.
+    pub ru_maxrss: u64,
+    /// The number of page faults serviced without any I/O activity;
+    /// here I/O activity is avoided by “reclaiming” a page frame from
+    /// the list of pages awaiting reallocation.
+    pub ru_minflt: u64,
+    /// The number of page faults serviced that required I/O activity.
+    pub ru_majflt: u64,
+    /// The number of times the filesystem had to perform input.
+    pub ru_inblock: u64,
+    /// The number of times the filesystem had to perform output.
+    pub ru_oublock: u64,
+    /// The number of times a context switch resulted due to a process
+    /// voluntarily giving up the processor before its time slice
+    /// was completed (usually to await availability of a
+    /// resource).
+    pub ru_nvcsw: u64,
+    /// The number of times a context switch resulted due to a higher
+    /// priority process becoming runnable or because the current
+    /// process exceeded its time slice.
+    pub ru_nivcsw: u64,
+}
+
+impl ResourceUsage {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub fn get() -> Self {
+        Self::from(getrusage())
+    }
+
+    pub fn runtime() -> Self {
+        let r = getrusage();
+        let mut u = ResourceUsage::new();
+
+        u.ru_utime = from_timeval(r.ru_utime).as_nanos() as u64;
+        u.ru_stime = from_timeval(r.ru_stime).as_nanos() as u64;
+        u.ru_nvcsw = r.ru_nvcsw as u64;
```

Comment:
if those are all the fields we need, I am wondering whether it wouldn't be best if we just don't declare the others?

If we want to declare all of them for some reason, maybe it is best to keep the ones we use together in the struct so they  fall in the same cacheline

Reply:
Wiped